### PR TITLE
Least-squares Fit

### PR DESCRIPTION
**Pull request by haiderabbas007**
# Pull Request

This PR introduces a directory with a complete pipeline for performing weighted least-squares linear fitting on grouped data from Table 1 of [Hubble's 1929 paper](https://www.pnas.org/doi/pdf/10.1073/pnas.15.3.168) along with visualizations and supporting documentation. The PR includes scripts, a data file, and a notebook to facilitate running the analysis.

### Files Added:

1) **grouped_data_generator.py:**

- Script to read, process, and group raw data by specific object labels.
- Calculates grouped averages and uncertainties, producing a filtered data summary saved as grouped_data_filtered.csv.

2) **linear_fit.py:**

- Contains functions _linear_fit_ and _read_data_from_csv_ to perform weighted least-squares fitting and load CSV data, respectively. These functions are used in the main fitting and plotting script. 
- _Author: dnxjay (github ID: 181004305)_

3) **fit_and_plot.py:**

- Script to read grouped and filtered data, apply the linear_fit function, and generate a plot.
- Displays grouped data with error bars and overlays the calculated best-fit line, including fit parameters.

4) **grouped_data_summary_filtered.csv:**

- Example output file from grouped_data_generator.py.
- Contains grouped, averaged values with uncertainties for use in fitting and plotting.

5) **how_to_run.ipynb:**

- Jupyter Notebook with instructions on running the scripts, including example outputs.
- Provides step-by-step guidance to replicate the analysis and visualization pipeline.

6) **raw_data_table1_grouped.md:**

- Markdown file with raw data in a tabular form.
- Used as input for grouped_data_generator.py to demonstrate the data processing pipeline.
- _Author: iglesias-cardinale (github ID: 112112966)_

### Overview of Workflow:

- **Data Processing:** grouped_data_generator.py processes raw data and outputs a filtered CSV for analysis.
- **Fitting and Plotting:** fit_and_plot.py performs linear fitting on the grouped data and visualizes the results.
- **Documentation:** how_to_run.ipynb provides detailed usage instructions.

### CODE REVIEW

**Comment by laserlab** [comment]
@ubsuny/cp1-24 Anyone who wants to review this?

**Comment by zbpetersbuf** [comment]
i can review this, when is this due?

**Comment by kylemasc917** [comment]
as soon as possible since this code should be merged by end of day today.

File: Least-squares Fit/raw_data_table1_grouped.md
Comment (zbpetersbuf) [inline Least-squares Fit/raw_data_table1_grouped.md:1]: compared this data to the pdf's, all good here

File: Least-squares Fit/how_to_run.ipynb
Comment (zbpetersbuf) [inline Least-squares Fit/how_to_run.ipynb:1]: im not sure how to review this but no issues i can see

**Review by zbpetersbuf — APPROVED** [review]
as far as I can tell this looks good

**Comment by zbpetersbuf** [comment]
the pytest error is from how there isnt a module named test_*.py

**Comment by zbpetersbuf** [comment]
the Flake8 error is from missing items: Fit/fit_and_plot.py, Fit/grouped_data_generator.py, Fit/linear_fit.py, and Least-squares however I also don't fully understand what Flake8 is so I could be wrong, please ignore both of my comments if not helpfull

**Comment by HeshamElsaman** [comment]
Great.
I guess I noticed a potential source for the error. I will try to solve it and run the test again before merging. If it's not solved, I'll merge anyways and continue searching for a potential solution.

**Comment by HeshamElsaman** [comment]
It seems that, although it's edited, it keeps running the old workflows when I get to re-run it.
Since it looks great to me as well, I'll be merging anyways.

**Pull request merged**